File: CS/SaveCheckStateTreeListNode(E3376)/SaveCheckStateTreeListNode/frmMain.cs
using System;
using System.Collections.Generic;
using System.Data;
using System.Linq;
using System.Windows.Forms;

namespace SaveCheckStateTreeListNode
{
    public partial class frmMain : Form
    {
        readonly DataTable source = new DataTable();
        const string resetTxt = "Reset";
        const string setupTxt = "Setup";
        const string nameCheckColumn = "Check";

        public frmMain()
        {
            InitializeComponent();
            FillTable();
            treeList1.DataSource = source;
            treeList1.KeyFieldName = "ID";
            treeList1.ParentFieldName = "ParentID";
            if (treeList1.DataSource != null)
                btnReset.Text = resetTxt;
            else
                btnReset.Text = setupTxt;
            treeList1.ExpandAll();
            new SaveCheckHelper(treeList1, nameCheckColumn);
        }

        void FillTable()
        {
            source.Columns.Add("ID");
            source.Columns.Add("ParentID");
            source.Columns.Add("FirstName");
            source.Columns.Add("LastName");
            source.Columns.Add("Age");
            source.Columns.Add("Address");
            source.Columns.Add(nameCheckColumn, typeof(CheckState));
            Random rnd = new Random();
            const int countRows = 20;
            for (int i = 0; i < countRows; i++)
            {
                source.Rows.Add(new object[] { i, rnd.Next(countRows - 1), "FirstName" + i, "LastName" + i, i + 20, "Address" + i, CheckState.Unchecked });
            }
        }

        private void btnReset_Click(object sender, EventArgs e)
        {
            if (btnReset.Text == resetTxt)
            {
                treeList1.DataSource = null;
                btnReset.Text = setupTxt;
            }
            else
            {
                treeList1.DataSource = source;
                btnReset.Text = resetTxt;
            }
            treeList1.ExpandAll();
        }
    }
}

File: VB/SaveCheckStateTreeListNode/frmMain.vb
Imports System
Imports System.Collections.Generic
Imports System.Data
Imports System.Linq
Imports System.Windows.Forms

Namespace SaveCheckStateTreeListNode
    Partial Public Class frmMain
        Inherits Form

        Private ReadOnly source As New DataTable()
        Private Const resetTxt As String = "Reset"
        Private Const setupTxt As String = "Setup"
        Private Const nameCheckColumn As String = "Check"

        Public Sub New()
            InitializeComponent()
            FillTable()
            treeList1.DataSource = source
            treeList1.KeyFieldName = "ID"
            treeList1.ParentFieldName = "ParentID"
            If treeList1.DataSource IsNot Nothing Then
                btnReset.Text = resetTxt
            Else
                btnReset.Text = setupTxt
            End If
            treeList1.ExpandAll()
            Dim tempVar As New SaveCheckHelper(treeList1, nameCheckColumn)
        End Sub

        Private Sub FillTable()
            source.Columns.Add("ID")
            source.Columns.Add("ParentID")
            source.Columns.Add("FirstName")
            source.Columns.Add("LastName")
            source.Columns.Add("Age")
            source.Columns.Add("Address")
            source.Columns.Add(nameCheckColumn, GetType(CheckState))
            Dim rnd As New Random()
            Const countRows As Integer = 20
            For i As Integer = 0 To countRows - 1
                source.Rows.Add(New Object() { i, rnd.Next(countRows - 1), "FirstName" & i, "LastName" & i, i + 20, "Address" & i, CheckState.Unchecked })
            Next i
        End Sub

        Private Sub btnReset_Click(ByVal sender As Object, ByVal e As EventArgs) Handles btnReset.Click
            If btnReset.Text = resetTxt Then
                treeList1.DataSource = Nothing
                btnReset.Text = setupTxt
            Else
                treeList1.DataSource = source
                btnReset.Text = resetTxt
            End If
            treeList1.ExpandAll()
        End Sub
    End Class
End Namespace
